### DISCUSSION TITLE
How to get public key from card with EC key

### DISCUSSION THREAD

**Discussion by msetina** [OpenSC/OpenSC#3088] [Q&A] [answered]
I have tried to get public key from a card with EC key (EOI driver). I used CKA_VALUE attribute to get bytes in DER format as I do for RSA key. The value I got I put  to cryptography API load_der_public_key, but it raised an error. The DER starts with 4 (OCTET STRING), the function expects a SEQUENCE.
What would be the proper way to get public key from a card with EC key?

**Comment by msetina** [comment]
After some searching and learning on the internet I found out that x,y point is considered public key. and I can confirm that the value I get from CKA_VALUE is uncompressed DER encoded EC point. The error I got from cryptography load_der_public_key call had all the info. I do not know what cryptography expects, but in case of xmldsig EC curve public key is described by x,y and curve name. So i understand that they can not make an instance of public key as som information is missing. Should DER length be enough get curve name? IDK

**Comment by popovec** [comment]
> Should DER length be enough get curve name?  

No, for example,  DER for secp256k1 and secp256r1 length is same, but another parameters of these curves are different. And vice versa, for the same curve, a point can be written with a different DER length (for example, the initial 04 in the DER record before the first coordinate indicates an uncompressed format, a compressed format allows you to omit a kind of coordinate (it can be calculated))

**Comment by msetina** [comment]
So guys from cryptography should explicitly define what DER encoded format they want in that function. For now I still can not get their public key class even with:
```EllipticCurvePublicKey.from_encoded_point(curve, ec_point)```
I got the right curve as their class, but they do not want ec_point in DER format I have. The python code does check that it starts with 4 which is a good start, but then they put it to openssl rust call to:
```openssl::ec::EcPoint::from_bytes```
which does not like the data or something.

**Comment by popovec** [comment]
> the function expects a SEQUENCE

DER encoding for sequence:
0x30, length, point

point encoding :
0x02, length, coordinate 1
0x02, length, coordinate 2

If  highest bit of coordinate is set to 1,  then 0x00 is prepended to this coordinate. Indicator 0x04 as uncompressed indicator is not used in this case

In the case of an EC curve, the point must lie on the curve - you can easily verify this by substituting the coordinates of the point into the equation of the curve.

**Reply by msetina** [reply]
Thank you. I am gettting EC point from a card with CKA_EC_POINT and it has a 0x04 at the start. It has 99 bytes. I am searching for info how to decode that.

**Reply by msetina** [reply]
It is strange: 
First 4 bytes are:
04610453 (4,97,4,83)
if this was an octet string containg octet string that would imply there are 12 bytes extra at the end.
On the other hand if it is only one octet string of 97 bytes length and as there is an FF ( 255) as last byte. In my case I need two 48 byte numbers for 384bit EC key. 
Would this corespond to any known formats?

**Comment by popovec** [comment, marked as answer]
TAG: 04 (octet string)
LENGTH: 0x61  (97 bytes)
VALUE: 04 ||  48 bytes for X || 48 bytes for Y coordinate.

I would try to put the 48 byte coordinates into the calculator and check if they match the secp384r1

**Reply by msetina** [reply]
Thank you. This format checks out with the formula. I think openssl does not know of this format.

**Reply by msetina** [reply]
What would value 04 on the third place mean? Something NXP knows and doesn't tell or is there some domain it coresponds to?

**Reply by popovec** [reply]
> What would value 04 on the third place mean?

These codes are not defined by ASN 1 coding but are defined in ANSI X9.62

04 - the following bytes represent the X and Y coordinate of the point
02,03  - the following bytes represent the compressed point coordinate (please use google to find decompression and compression methods - without curve parameters, this point cannot be decompressed/compressed.)

In this place, the code can also represent "point of infinity"  (0x00) and other hybrid coding methods.

**Comment by dengert** [comment]
To better understand ASN.1, search the internet for "layman's guide to asn1" from 1993 It can be found in several formats. 

The EC uncompressed format, as pointed out above is 04||x||y where x,y  are big endian integers both the same size. There is a 1 in 2^16 chance they both may start with a zero byte, in which  x and y may be 23 bytes each and not 24. (And a 1 in 2^64 that two bytes are zero in both.)

Note that the 04||x||y is not ASN.1 and x and y may be returned from a card as separate integers.  I have seen cards that will return 
x and or y without leading zeros.  The uncompressed format can have leading zeros as it is not ASN.1 

As you have discovered, smartcards and crypto follow many standards some times, but many card vendors have done things on their own and not followed any standards. They usually also provide some card driver for their cards for Windows only, and little or no documentation on the card commands. 

Most cards  that support EC only support a limited number of curves. 
In OpenSC source    `src/libopensc/`  grep for _sc_card_add_ec_alg  which takes a  curve size and curve OID. 

card-eoi.c only supports one curve (as far as we know) and the comment is also helpful.
```
static struct {
        int len;
        struct sc_object_id oid;
} eoi_curves[] = {
        /* secp384r1 */
        {384, {{1, 3, 132, 0, 34, -1}}}
};
```
The EOI driver is new, and says it require openpace and CAN, which introduces two more standards, which I known very little about. 

So your quest to get a public key from the card is dependent on PKCS11, ANSI X9.62, PKCS15, openpace, CAN, ISO7816-4 and maybe others and vendor specific code. 

https://docs.oasis-open.org/pkcs11/pkcs11-curr/v3.0/os/pkcs11-curr-v3.0-os.pdf  Table 34, Elliptic Curve Public Key Object Attributes defines CKA_EC_POINT.   and see  https://github.com/OpenSC/OpenSC/issues/3000 as we may have gotten it wrong.

**Comment by msetina** [comment]
Thank you for this information. I would appreciate a heads up when #3000 is resolved as it will probably influence the result of this saga of mine. 
I will prepare a OpenPGP card with EC key to see how to best open the resolution to be able to survive different implementations (big words)
I was hoping that this is already packaged somewhere (cryptography or ), but now I know more.
I am surprised that openssl does not have this encoding support because if I read correctly ANSI X9.62 is not an unknown for this type of data.

**Comment by dengert** [comment]
ANSI X9.62 goes back to standards for banking.  ASN.1 goes back to telephones.  Official documentation for both is for sale, but you might find some versions on the internet. 

https://neuromancer.sk/std/ might be helpful, as is any IETF RFCs dealing with ECDSA or ECDH. https://datatracker.ietf.org/doc/rfc8410/ which describes some curves that are similar to EC and use PKCS11 CKA_EC_PARAMS.    

I need to get back to: ~https://github.com/dengert/OpenSC/tree/openpgp-ec-pub-curve~  https://github.com/dengert/OpenSC/tree/X25519-improvements-2 which deals with a lot of the issues of mapping curves and better integration for the X25519 curves in OpenPGP in OpenSC.
